assets/js/app/LocusZoom.js

```diff
@@ -319,42 +319,54 @@ LocusZoom.parseFields = function (data, html) {
-        var then = x[1];
-        var rest = x[2];
-        if ((new LocusZoom.Data.Field(condition)).resolve(data)) {
-            html = before + then + rest;
-        } else {
-            html = before + rest;
-        }
-    }
-    // Match all things that look like fields in the HTML
-    var matches = html.match(/{{[A-Za-z0-9_:|]+}}/g);
-    if (matches){
-        // Remove duplicates
-        matches.reduce(function(a,b){if(a.indexOf(b)<0)a.push(b);return a;},[]);
-        // Replace matches with resolved values from the data object
-        matches.forEach(function(match){
-            var field = new LocusZoom.Data.Field(match.substring(2, match.length-2));
-            var value = field.resolve(data);
-            if (["string","number","boolean"].indexOf(typeof value) != -1){
-                html = html.replace(match, value);
-            } else if (value == null){
-                html = html.replace(match, "");
-            }
-        });
+    // `tokens` is like [token,...]
+    // `token` is like {text: '...'} or {variable: 'foo|bar'} or {condition: 'foo|bar'} or {close: 'if'}
+    var tokens = [];
+    var regex = /\{\{(?:(#if )?([A-Za-z0-9_:\|]+)|(\/if))\}\}/;
+    while (html.length > 0){
+        var m = regex.exec(html);
+        if (!m) { tokens.push({text: html}); html = ''; }
+        else if (m.index != 0) { tokens.push({text: html.slice(0, m.index)}); html = html.slice(m.index); }
+        else if (m[1] == "#if ") { tokens.push({condition: m[2]}); html = html.slice(m[0].length); }
+        else if (m[2]) { tokens.push({variable: m[2]}); html = html.slice(m[0].length); }
+        else if (m[3] == "/if") { tokens.push({close: 'if'}); html = html.slice(m[0].length); }
+        else { throw [m, html] }
     }
-    return html;
+    var astify = function() {
+        var token = tokens.shift();
+        if (token.text || token.variable) {
+            return token;
+        } else if (token.condition) {
+            token.then = [];
+            while(tokens[0].close != 'if') token.then.push(astify());
+            tokens.shift(); //consume {{/if}}
+            return token;
+        } else { throw 'what?' }
+    };
+    // `ast` is like [thing,...]
+    // `thing` is like {text: '...'} or {variable:'foo|bar'} or {condition: 'foo|bar', then:[thing,...]}
+    var ast = [];
+    while (tokens.length > 0) ast.push(astify());
+
+    var resolve = function(variable) {
+        if (!resolve.cache.hasOwnProperty(variable))
+            resolve.cache[variable] = (new LocusZoom.Data.Field(variable)).resolve(data);
+        return resolve.cache[variable];
+    };
+    resolve.cache = {};
+    var render_node = function(node) {
+        if (node.text) {
+            return node.text;
+        } else if (node.variable) {
+            var value = resolve(node.variable);
+            return (["string","number","boolean"].indexOf(typeof value) != -1) ? value :
```

Comment:
I've always loved the way that order-of-operations lets you write switch-case-like nested ternaries.  Anyways, I've changed it.